src/redis-cli.c

```diff
@@ -418,91 +425,226 @@ static sds cliVersion(void) {
-        helpEntries[pos++] = tmp;
-    }
-}
-
-/* cliInitHelp() setups the helpEntries array with the command and group
- * names from the help.h file. However the Redis instance we are connecting
- * to may support more commands, so this function integrates the previous
- * entries with additional entries obtained using the COMMAND command
- * available in recent versions of Redis. */
-static void cliIntegrateHelp(void) {
+    int nextIndex = 0;
     if (cliConnect(CC_QUIET) == REDIS_ERR) return;
 
     redisReply *reply = redisCommand(context, "COMMAND");
     if(reply == NULL || reply->type != REDIS_REPLY_ARRAY) return;
 
-    /* Scan the array reported by COMMAND and fill only the entries that
-     * don't already match what we have. */
+    /* Scan the array reported by COMMAND and fill in the entries */
+    helpEntriesLen = cliCountCommands(reply);
+    helpEntries = zmalloc(sizeof(helpEntry)*helpEntriesLen);
+
     for (size_t j = 0; j < reply->elements; j++) {
-        redisReply *entry = reply->element[j];
-        if (entry->type != REDIS_REPLY_ARRAY || entry->elements < 4 ||
-            entry->element[0]->type != REDIS_REPLY_STRING ||
-            entry->element[1]->type != REDIS_REPLY_INTEGER ||
-            entry->element[3]->type != REDIS_REPLY_INTEGER) return;
-        char *cmdname = entry->element[0]->str;
-        int i;
-
-        for (i = 0; i < helpEntriesLen; i++) {
-            helpEntry *he = helpEntries+i;
-            if (!strcasecmp(he->argv[0],cmdname))
-                break;
-        }
-        if (i != helpEntriesLen) continue;
-
-        helpEntriesLen++;
-        helpEntries = zrealloc(helpEntries,sizeof(helpEntry)*helpEntriesLen);
-        helpEntry *new = helpEntries+(helpEntriesLen-1);
-
-        new->argc = 1;
-        new->argv = zmalloc(sizeof(sds));
-        new->argv[0] = sdsnew(cmdname);
-        new->full = new->argv[0];
-        new->type = CLI_HELP_COMMAND;
-        sdstoupper(new->argv[0]);
-
-        struct commandHelp *ch = zmalloc(sizeof(*ch));
-        ch->name = new->argv[0];
-        ch->params = sdsempty();
-        int args = llabs(entry->element[1]->integer);
-        args--; /* Remove the command name itself. */
-        if (entry->element[3]->integer == 1) {
-            ch->params = sdscat(ch->params,"key ");
-            args--;
-        }
-        while(args-- > 0) ch->params = sdscat(ch->params,"arg ");
-        if (entry->element[1]->integer < 0)
-            ch->params = sdscat(ch->params,"...options...");
-        ch->summary = "Help not available";
-        ch->group = 0;
-        ch->since = "not known";
-        new->org = ch;
-    }
+        redisReply *command = reply->element[j];
+        if (command->type != REDIS_REPLY_ARRAY || command->elements < 4 ||
+            command->element[0]->type != REDIS_REPLY_STRING ||
+            command->element[1]->type != REDIS_REPLY_INTEGER ||
+            command->element[3]->type != REDIS_REPLY_INTEGER) return;
```

Comment:
what are these checks?
please add a comment, but also, i couldn't find that we used these, and maybe we should check others?

Reply:
These checks were in the original code. Apparently they're some kind of sanity check on the output of COMMAND? It's not clear to me why these specific checks appeared but not others.

Reply:
since we're revamping this area, i think we should take this opportunity to change this, or document what exactly it's for..

Reply:
Okay, these checks were added in b6ab4d0: "Fix redis-cli rare crash. This happens if the server (mysteriously) returns an unexpected response to the COMMAND command."

Basically, the code checks that each of the fields it intends to read from the COMMAND reply array actually exists and has the correct expected type. If not, it returns early with no error message.

If we want to maintain the equivalent behaviour, these checks should be removed (except for `element[0]`), since the new code doesn't read the same fields, and the command parsing code in `cliInitCommandHelpEntry()` should check the validity of each field before reading it.